crypto/bio/bio_lib.c

```diff
@@ -94,12 +94,21 @@ int BIO_set(BIO *bio, BIO_METHOD *method)
     bio->num_read = 0L;
     bio->num_write = 0L;
     CRYPTO_new_ex_data(CRYPTO_EX_INDEX_BIO, bio, &bio->ex_data);
-    if (method->create != NULL)
+
+    bio->lock = CRYPTO_THREAD_lock_new();
+    if (bio->lock == NULL) {
+        BIO_free(bio);
+        return 0;
+    }
+
+    if (method->create != NULL) {
         if (!method->create(bio)) {
-            CRYPTO_free_ex_data(CRYPTO_EX_INDEX_BIO, bio, &bio->ex_data);
-            return (0);
+            BIO_free(bio);
```

Comment:
If BIO_set is invoked by BIO_new; doing BIO_free(bio) here and returning 0 will cause to do OPENSSL_free (bio) twice. First in BIO_free(bio) and second in BIO_new() [ when it sees BIO_set failed, it is invoking OPENSSL_free()]